File: soungmunkim/Python/Graph/Graph_DFS_withColors.py
"""
[dfs로 graph 구현하기]

(base) soungmunkim@gimseongmun-ui-MacBookPro Graph % python -i Graph_DFS_withColors.py
>>> u, v, w = GNode("u"), GNode("v"), GNode("w")
>>> x, y, z = GNode("x"), GNode("y"), GNode("z")
>>> G = dict()
>>> G[u], G[v], G[w], G[x], G[y], G[z] = [v,x], [y], [y,z], [v], [x], [z]
>>> printGraph(G)
(u, W, (-1, -1), None): (v, W, (-1, -1), None)(x, W, (-1, -1), None)
(v, W, (-1, -1), None): (y, W, (-1, -1), None)
(w, W, (-1, -1), None): (y, W, (-1, -1), None)(z, W, (-1, -1), None)
(x, W, (-1, -1), None): (v, W, (-1, -1), None)
(y, W, (-1, -1), None): (x, W, (-1, -1), None)
(z, W, (-1, -1), None): (z, W, (-1, -1), None)
>>> 
>>> dfs(G)
=========================================
(u, W, (-1, -1), None): (v, W, (-1, -1), None)(x, W, (-1, -1), None)
(v, W, (-1, -1), None): (y, W, (-1, -1), None)
(w, W, (-1, -1), None): (y, W, (-1, -1), None)(z, W, (-1, -1), None)
(x, W, (-1, -1), None): (v, W, (-1, -1), None)
(y, W, (-1, -1), None): (x, W, (-1, -1), None)
(z, W, (-1, -1), None): (z, W, (-1, -1), None)
=========================================
(u, G, (1, -1), None): (v, W, (-1, -1), None)(x, W, (-1, -1), None)
(v, W, (-1, -1), None): (y, W, (-1, -1), None)
(w, W, (-1, -1), None): (y, W, (-1, -1), None)(z, W, (-1, -1), None)
(x, W, (-1, -1), None): (v, W, (-1, -1), None)
(y, W, (-1, -1), None): (x, W, (-1, -1), None)
(z, W, (-1, -1), None): (z, W, (-1, -1), None)
=========================================
(u, G, (1, -1), None): (v, G, (2, -1), u)(x, W, (-1, -1), None)
(v, G, (2, -1), u): (y, W, (-1, -1), None)
(w, W, (-1, -1), None): (y, W, (-1, -1), None)(z, W, (-1, -1), None)
(x, W, (-1, -1), None): (v, G, (2, -1), u)
(y, W, (-1, -1), None): (x, W, (-1, -1), None)
(z, W, (-1, -1), None): (z, W, (-1, -1), None)
=========================================
(u, G, (1, -1), None): (v, G, (2, -1), u)(x, W, (-1, -1), None)
(v, G, (2, -1), u): (y, G, (3, -1), v)
(w, W, (-1, -1), None): (y, G, (3, -1), v)(z, W, (-1, -1), None)
(x, W, (-1, -1), None): (v, G, (2, -1), u)
(y, G, (3, -1), v): (x, W, (-1, -1), None)
(z, W, (-1, -1), None): (z, W, (-1, -1), None)
=========================================
(u, G, (1, -1), None): (v, G, (2, -1), u)(x, G, (4, -1), y)
(v, G, (2, -1), u): (y, G, (3, -1), v)
(w, W, (-1, -1), None): (y, G, (3, -1), v)(z, W, (-1, -1), None)
(x, G, (4, -1), y): (v, G, (2, -1), u)
(y, G, (3, -1), v): (x, G, (4, -1), y)
(z, W, (-1, -1), None): (z, W, (-1, -1), None)
=========================================
(u, G, (1, -1), None): (v, G, (2, -1), u)(x, B, (4, 5), y)
(v, G, (2, -1), u): (y, G, (3, -1), v)
(w, W, (-1, -1), None): (y, G, (3, -1), v)(z, W, (-1, -1), None)
(x, B, (4, 5), y): (v, G, (2, -1), u)
(y, G, (3, -1), v): (x, B, (4, 5), y)
(z, W, (-1, -1), None): (z, W, (-1, -1), None)
=========================================
(u, G, (1, -1), None): (v, G, (2, -1), u)(x, B, (4, 5), y)
(v, G, (2, -1), u): (y, B, (3, 6), v)
(w, W, (-1, -1), None): (y, B, (3, 6), v)(z, W, (-1, -1), None)
(x, B, (4, 5), y): (v, G, (2, -1), u)
(y, B, (3, 6), v): (x, B, (4, 5), y)
(z, W, (-1, -1), None): (z, W, (-1, -1), None)
=========================================
(u, G, (1, -1), None): (v, B, (2, 7), u)(x, B, (4, 5), y)
(v, B, (2, 7), u): (y, B, (3, 6), v)
(w, W, (-1, -1), None): (y, B, (3, 6), v)(z, W, (-1, -1), None)
(x, B, (4, 5), y): (v, B, (2, 7), u)
(y, B, (3, 6), v): (x, B, (4, 5), y)
(z, W, (-1, -1), None): (z, W, (-1, -1), None)
=========================================
(u, B, (1, 8), None): (v, B, (2, 7), u)(x, B, (4, 5), y)
(v, B, (2, 7), u): (y, B, (3, 6), v)
(w, W, (-1, -1), None): (y, B, (3, 6), v)(z, W, (-1, -1), None)
(x, B, (4, 5), y): (v, B, (2, 7), u)
(y, B, (3, 6), v): (x, B, (4, 5), y)
(z, W, (-1, -1), None): (z, W, (-1, -1), None)
=========================================
(u, B, (1, 8), None): (v, B, (2, 7), u)(x, B, (4, 5), y)
(v, B, (2, 7), u): (y, B, (3, 6), v)
(w, G, (9, -1), None): (y, B, (3, 6), v)(z, W, (-1, -1), None)
(x, B, (4, 5), y): (v, B, (2, 7), u)
(y, B, (3, 6), v): (x, B, (4, 5), y)
(z, W, (-1, -1), None): (z, W, (-1, -1), None)
=========================================
(u, B, (1, 8), None): (v, B, (2, 7), u)(x, B, (4, 5), y)
(v, B, (2, 7), u): (y, B, (3, 6), v)
(w, G, (9, -1), None): (y, B, (3, 6), v)(z, G, (10, -1), w)
(x, B, (4, 5), y): (v, B, (2, 7), u)
(y, B, (3, 6), v): (x, B, (4, 5), y)
(z, G, (10, -1), w): (z, G, (10, -1), w)
=========================================
(u, B, (1, 8), None): (v, B, (2, 7), u)(x, B, (4, 5), y)
(v, B, (2, 7), u): (y, B, (3, 6), v)
(w, G, (9, -1), None): (y, B, (3, 6), v)(z, B, (10, 11), w)
(x, B, (4, 5), y): (v, B, (2, 7), u)
(y, B, (3, 6), v): (x, B, (4, 5), y)
(z, B, (10, 11), w): (z, B, (10, 11), w)
=========================================
(u, B, (1, 8), None): (v, B, (2, 7), u)(x, B, (4, 5), y)
(v, B, (2, 7), u): (y, B, (3, 6), v)
(w, B, (9, 12), None): (y, B, (3, 6), v)(z, B, (10, 11), w)
(x, B, (4, 5), y): (v, B, (2, 7), u)
(y, B, (3, 6), v): (x, B, (4, 5), y)
(z, B, (10, 11), w): (z, B, (10, 11), w)
"""

import collections

class GNode:
    def __init__(self, id, c = 'W', start = -1, finish = -1, p = None):
        # 그래프 노드의 초기화
        self.id = id         # 노드의 고유 ID
        self.color = c       # 노드의 색상 ('W': 흰색, 'G': 회색, 'B': 검은색)
        self.start = start    # 시작 노드로부터의 시작 시간
        self.finish = finish           # 노드의 종료 시간
        self.parent = p      # 부모 노드

    def __str__(self):
        # 노드의 문자열 표현을 반환
        if (self.parent != None):
            return "(" + self.id + ", " + self.color + ", (" \
                + str(self.start) + ", " + str(self.finish) + "), " + self.parent.id + ")"
        else:
            return "(" + self.id + ", " + self.color + ", (" \
                + str(self.start) + ", " + str(self.finish) + "), None)"

def printGraph(g):
    # 그래프를 화면에 출력
    for key, val in g.items():
        print(str(key) + ": ", end="")
        for u in val:
            print(str(u), end="")
        print("")

def dfs(G):
    time = 0  # 탐색 시간을 나타내는 변수

    def printG(d):
        # 그래프 출력용 도우미 함수
        print("=========================================")
        for key, val in d.items():
            print(str(key) + ": ", end="")
            for u in val:
                print(str(u), end="")
            print("")

    def dfs_visit(G, u):
        nonlocal time   # 외부 함수의 변수를 참조
        time = time + 1
        u.start = time  # 시작 시간 설정
        u.color = "G"      # 노드의 색상을 회색으로 변경
        printG(G)
        
        for v in G[u]:  # u의 인접 노드들에 대해
            if (v.color == "W"):  # 만약 노드 v가 방문되지 않았다면
                v.parent = u     # 부모 설정
                dfs_visit(G, v)  # 재귀적으로 탐색
        u.color = "B"  # 모든 인접 노드를 방문한 후, 노드의 색상을 검은색으로 변경
        time = time + 1
        u.finish = time  # 종료 시간 설정
        printG(G)
        
    for key in G:
        key.color = "W"  # 모든 노드를 흰색으로 초기화
        key.start = -1
        key.finish = -1
        key.parent = None
    printG(G)
    
    for u in G:  # 모든 노드에 대해
        if (u.color == "W"):  # 방문되지 않은 노드라면
            dfs_visit(G, u)  # 깊이 우선 탐색 시작

            

File: soungmunkim/C/Graph/Graph_DFS_withColors.c
/*
[dfs로 graph 구현하기 with colors]
*/

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

#define MAX_NODES 100  // 그래프의 최대 노드 수

// 그래프 노드를 표현하기 위한 구조체
typedef struct GNode {
    char id[10];            // 노드의 고유 ID (문자열 형태)
    char color;             // 노드의 색상 ('W': 흰색, 'G': 회색, 'B': 검은색)
    int start, finish;      // DFS 탐색에서의 시작 시간과 종료 시간
    struct GNode* parent;   // 부모 노드의 포인터
} GNode;

// 그래프를 표현하기 위한 구조체
typedef struct {
    GNode* nodes[MAX_NODES];               // 그래프의 모든 노드를 저장하는 배열
    GNode* adjList[MAX_NODES][MAX_NODES];  // 인접 리스트
    int nodeCount;                         // 그래프의 노드 수
} Graph;

Graph graph;  // 전역 그래프 변수
int time_count = 0;  // DFS 탐색에서의 전체 시간을 계산하는 변수

// 그래프의 노드 정보를 출력하는 함수
void printNode(GNode* node) {
    printf("(%s, %c, (%d, %d), ", node->id, node->color, node->start, node->finish);
    if (node->parent) {
        printf("%s) ", node->parent->id);
    } else {
        printf("None) ");
    }
}

// 그래프의 정보를 출력하는 함수
void printGraph() {
    for (int i = 0; i < graph.nodeCount; i++) {
        printNode(graph.nodes[i]);
        printf(": ");
        for (int j = 0; graph.adjList[i][j] != NULL; j++) {
            printNode(graph.adjList[i][j]);
        }
        printf("\n");
    }
}

// DFS 탐색의 핵심 함수
void dfs_visit(GNode* u) {
    time_count++;
    u->start = time_count;
    u->color = 'G';

    printGraph();

    for (int i = 0; graph.adjList[u->id[0] - 'u'][i] != NULL; i++) {
        GNode* v = graph.adjList[u->id[0] - 'u'][i];
        if (v->color == 'W') {
            v->parent = u;
            dfs_visit(v);
        }
    }

    u->color = 'B';
    time_count++;
    u->finish = time_count;

    printGraph();
}

// 모든 노드에 대해 깊이 우선 탐색을 시작하는 함수
void dfs() {
    for (int i = 0; i < graph.nodeCount; i++) {
        graph.nodes[i]->color = 'W';
        graph.nodes[i]->start = -1;
        graph.nodes[i]->finish = -1;
        graph.nodes[i]->parent = NULL;
    }

    printGraph();

    for (int i = 0; i < graph.nodeCount; i++) {
        GNode* u = graph.nodes[i];
        if (u->color == 'W') {
            dfs_visit(u);
        }
    }
}

int main() {
    GNode u = {"u"};
    GNode v = {"v"};
    GNode w = {"w"};
    GNode x = {"x"};
    GNode y = {"y"};
    GNode z = {"z"};

    graph.nodes[0] = &u;
    graph.nodes[1] = &v;
    graph.nodes[2] = &w;
    graph.nodes[3] = &x;
    graph.nodes[4] = &y;
    graph.nodes[5] = &z;
    graph.nodeCount = 6;

    graph.adjList[0][0] = &v;
    graph.adjList[0][1] = &x;

    graph.adjList[1][0] = &y;

    graph.adjList[2][0] = &y;
    graph.adjList[2][1] = &z;

    graph.adjList[3][0] = &v;

    graph.adjList[4][0] = &x;

    graph.adjList[5][0] = &z;

    printGraph();
    dfs();

    return 0;
}
